### DISCUSSION TITLE
Modularity in Theia

### DISCUSSION THREAD

**Discussion by imaheshraj** [eclipse-theia/theia#11793] [General]
Hi,
I was going thru Theia's source code and found that most Theia's components are modularized.
Like other components are dependent on @theia/core and not vice versa. 

I believe inversify helps in this regard. Apart from that is there any package I can check on to learn about this. 

It would be of great if you could point me in direction to learn more about this. 

Thanks

**Comment by vince-fugnitto** [comment]
@imaheshraj thank you for the discussion!

Theia uses `inversify` as a form of dependency inversion in the framework, which in turns offers downstream applications the ability to extend and re-implement core features as their use-cases require. You can learn a bit more regarding `inversify` in the project by taking at look at our documentation (https://theia-ide.org/docs/services_and_contributions/#dependency-injection-di) and following the blog post linked (https://eclipsesource.com/blogs/2018/11/28/how-to-inversify-in-eclipse-theia/)

> I was going thru Theia's source code and found that most Theia's components are modularized. Like other components are dependent on @theia/core and not vice versa.

That is true, `@theia/core` is the base of the entire framework and the extension from which other extensions build upon. We keep extensions concise in their functionality and/or contributions and try to not couple them too much or unnecessarily. The decoupling/modularity means that downstream applications can pick and choose what features they want more easily, but there is always ways to remove functionality with custom extensions if needed. Note that with the inclusion of vscode plugin support, a lot of extensions will be pulled as a result.
